### PR TITLE
Fix for builing DBD-mysql together with MariaDB Connector/C.

### PR DESCRIPTION
**Pull request by 9EOR9**
Use mysql_option function instead of accessing internal members of MYSQL structure.

### CODE REVIEW

**Comment by pali** [comment]
Some tests failed:
```
dbdimp.c:2272:29: error: ‘MYSQL_OPT_RECONNECT’ undeclared (first use in this function)
```

Also note that MariaDB Connector/C does not work with DBD::mysql, because MariaDB Connector/C has broken zerofill support, see: https://github.com/perl5-dbi/DBD-mysql/commit/e1ea5521675759634ad9e2e83118fd721d679a26 https://github.com/perl5-dbi/DBD-mysql/pull/98#issuecomment-278932984

**Comment by pali** [comment]
This could help you:

http://web.archive.org/web/20110902122038/http://dev.mysql.com/doc/refman/5.0/en/auto-reconnect.html
> The MYSQL_OPT_RECONNECT option is available as of MySQL 5.0.13.

File: dbdimp.h [hunk 1]
```diff
@@ -110,22 +124,22 @@
  */
 
 /* Use mysql_options with MYSQL_OPT_SSL_VERIFY_SERVER_CERT */
-#if ((MYSQL_VERSION_ID >= 50023 && MYSQL_VERSION_ID < 50100) || MYSQL_VERSION_ID >= 50111) && (MYSQL_VERSION_ID < 80000 || defined(MARIADB_BASE_VERSION))
+#ifdef CLIENT_SSL_VERIFY_SERVER_CERT
```
Comment (9EOR9) [inline dbdimp.h]: The verification is still in place:
https://github.com/mysql/mysql-server/blob/8.0/sql-common/client.cc#L3715

Even if MYSQL_OPT_SSL_VERIFY_SERVER_CERT was renoved, you can still pass CLIENT_SSL_VERIFY_SERVER_CERT flag to mysql_real_connect (last parameter).

Comment (9EOR9) [inline dbdimp.h]: The verification is still in place:
https://github.com/mysql/mysql-server/blob/8.0/sql-common/client.cc#L3715

Even if `MYSQL_OPT_SSL_VERIFY_SERVER_CERT` was renoved, you can still pass CLIENT_SSL_VERIFY_SERVER_CERT flag to mysql_real_connect (last parameter).

Comment (pali) [inline dbdimp.h]: But this check is for defining `HAVE_SSL_VERIFY`. And it means DBD::mysql can use mysql_options() with `MYSQL_OPT_SSL_VERIFY_SERVER_CERT`. It has nothing with `CLIENT_SSL_VERIFY_SERVER_CERT` flag or mysql_real_connect() function.

Some of those mysql_* SSL related functions are broken in particular MySQL versions and therefore we need to correctly ifdef them and use safe variant.

See https://github.com/perl5-dbi/DBD-mysql/pull/114 and https://github.com/perl5-dbi/DBD-mysql/issues/110

Your change break SSL verification for some MySQL versions.

Comment (pali) [inline dbdimp.h]: So... I guess correct should be this check which do not break SSL verification:
```
#if ((MYSQL_VERSION_ID >= 50023 && MYSQL_VERSION_ID < 50100) || MYSQL_VERSION_ID >= 50111) && (MYSQL_VERSION_ID < 80000 || defined(MARIADB_CLIENT))
```

File: dbdimp.c [hunk 2]
```diff
@@ -2268,7 +2269,7 @@ MYSQL *mysql_dr_connect(
         we turn off Mysql's auto reconnect and handle re-connecting ourselves
         so that we can keep track of when this happens.
       */
-      result->reconnect=0;
```
Comment (pali) [inline dbdimp.c]:
```
#if MYSQL_VERSION_ID >= 50013
mysql_options(result, MYSQL_OPT_RECONNECT, &reconnect);
#else
result->reconnect=0;
#endif
```

File: dbdimp.h (hunk 1 above)
Comment (9EOR9) [inline dbdimp.h]: Yes, this looks better, I will fix and push later today. We need to think about to add ssl_mode also in Connector/C, so API's will not differ too much and iit would be easier to handle TLS/SSL options in the future (e.g. Connector/C additionally supports finger print check of server certificate).

**Comment by AdamWill** [comment]
Note downstream (Fedora) bug: https://bugzilla.redhat.com/show_bug.cgi?id=1470196 , where one of our mariadb maintainers suggests adding `+#include <mysql_version.h>  /* For MYSQL_VERSION_ID */` for the versioning stuff...

**Comment by pali** [comment]
@AdamWill that would not work, because `MYSQL_VERSION_ID` was added only recently in commit MariaDB Connector/C commit: https://github.com/MariaDB/mariadb-connector-c/commit/422d0f786e5e6eb103000a75738322c8baa9aa12

**Comment by AdamWill** [comment]
@pali but no stable version was ever actually released without it, AFAICT. On the 2.x branch, nothing was renamed to `MARIADB`, so it still had `MYSQL_VERSION_ID`, see e.g. the 2.3.2 tag:

https://github.com/MariaDB/mariadb-connector-c/blob/v2.3.2/include/mysql_version.h.in

and in 3.0.2, the commit you mention was already included:

https://github.com/MariaDB/mariadb-connector-c/blob/v3.0.2/include/mariadb_version.h.in

So the only tagged release that does not have `MYSQL_VERSION_ID` is 3.0.1-beta, which was not a stable release. Does this still concern you?

**Comment by pali** [comment]
Ou right, so problematic is only the 3.0.1-beta and all older versions are working... Then we can use it.

**Comment by AdamWill** [comment]
Well, see my edit (the file still seems to have a different name on the 3.x branch). Also note the 'downstream bug' I referenced isn't actually for quite the same thing, it's about building against MariaDB 10.2 vs. 10.1 (not building against mariadb-connector-c).

Actually the more I look at this 3.x branch of mariadb-connector-c, the weirder it gets...AIUI, mariadb-connector-c / mysql-connector-c are supposed to be just the client library from the 'main' distribution (mariadb / mysql) isolated so you don't have to install the whole thing if you're just building a client. But mariadb-connector-c 3.x doesn't seem to match the client library from mariadb 10.2 at all, at least in terms of these header files. They haven't all been renamed in MariaDB itself - there's no `mariadb_` or `ma_` headers in https://github.com/MariaDB/server/tree/10.2/include . So that all seems a bit odd...

**Comment by 9EOR9** [comment]
Both MariaDB 10.2.7 (released last week) and Connector/C 3.0.2 (will be released today) have MYSQL_VERSION_ID

**Comment by 9EOR9** [comment]
Adam,
Connector/C 3.0 is identical to the client library in 10.2. Please note that Connector/C in server tree is a git submodule in directory libmariadb (Connector/C branch 10.2-server).

**Comment by AdamWill** [comment]
@9EOR9 But what about the header file names? They seem to be different between the two...

**Comment by 9EOR9** [comment]
What header file names do you mean? The only file which should be included is mysql.h - it includes all necessary files for client applications.

**Comment by AdamWill** [comment]
Hmm, I see. Well, the patch our maintainer suggested included an explicit `#include mysql_version.h` - so I guess that should be left out...

**Comment by AdamWill** [comment]
@9EOR9 So poking into this further, I found what seems to be a bug in MariaDB itself: when you run an install of MariaDB 10.2, you ultimately wind up with the `mysql.h` from the `libmariadb/` subdirectory (which, as you pointed out, is mariadb-connector-c as a subproject) installed. *Not* the `mysql.h` from MariaDB itself. So you get the `mysql.h` for the client library, which ultimately doesn't provide as much stuff as the 'full fat' `mysql.h` for the whole of MariaDB. This ultimately seems to be behind several of the build failures we observed with things that want to use stuff that's defined in MariaDB's `mysql_version.h`, which MariaDB's `mysql.h` includes, but *not* in the client library's `mariadb_version.h`, which the client library's `mysql.h` includes. Reported as https://jira.mariadb.org/browse/MDEV-13370 .

**Comment by GieltjE** [comment]
The bug has been fixed by MariaDB in 10.2.9, we really should be working on a fix as half the world is migrating to MariaDB and a lot of packages depend on this package (spamassassin, opendkim, opendmarc)

**Comment by noplanman** [comment]
Any update on this? Has just smacked me in the face 😕

**Comment by mannih** [comment]
FWIW: I just ran into this problem after upgrading my fedora install from 26 to 27. I was able to install DBD::mysql after applying this patch set: https://src.fedoraproject.org/rpms/perl-DBD-MySQL/blob/c9a8c6732a3cad7016fe0cbf217a9f0846af6bb9/f/DBD-mysql-4.043-Fix-build-failures-for-MariaDB-10.2.patch
Of course, I have no idea how well this works on other systems with other libraries.

**Comment by GieltjE** [comment]
@mannih Just tested it on debian with MariaDB 10.2.12 with spamassassin, opendkim and opendmarc, everything working as it should (also applied the other pathes availble in https://src.fedoraproject.org/rpms/perl-DBD-MySQL/tree/master).

**Comment by pali** [comment]
Slightly different approach was used to fix these problems in https://github.com/gooddata/DBD-MariaDB repository.

**Comment by mbeijen** [comment]
Partly fixed with 0b1884f4153d3de0a91bae4d3b3cc1771ffdf4f3

**Comment by dveeden** [comment]
As far as I'm aware this PR is not needed anymore. Please ping me if that is not the case, then we can re-open this PR and actually get the issue fixed.

**Pull request closed without merging**